src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs

```diff
@@ -0,0 +1,131 @@
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal class CodeActionEndpoint : ICodeActionHandler
+    {
+        private readonly IEnumerable<RazorCodeActionProvider> _providers;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly ILogger _logger;
+
+        private CodeActionCapability _capability;
+
+        public CodeActionEndpoint(
+            IEnumerable<RazorCodeActionProvider> providers,
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver,
+            ILoggerFactory loggerFactory)
+        {
+            if (providers is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _providers = providers;
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+            _logger = loggerFactory.CreateLogger<CodeActionEndpoint>();
+        }
+
+        public CodeActionRegistrationOptions GetRegistrationOptions()
+        {
+            return new CodeActionRegistrationOptions()
+            {
+                DocumentSelector = RazorDefaults.Selector
+            };
+        }
+
+        public async Task<CommandOrCodeActionContainer> Handle(CodeActionParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+
+            if (document is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
```

Comment:
is it appropriate in this domain to do:
```suggestion
            var codeDocument = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
```
?

Reply:
I've never seen this before, @NTaylorMullen @ajaybhargavb

Reply:
Yes;

FAQ @noahbkim: https://devblogs.microsoft.com/dotnet/configureawait-faq/

Reply:
Ya it's best practice; however, it's something we've failed to do in a lot of locations. Basically what it's saying is that once you resume the tasks execution that it doesn't have to try and re-delegate it to the same context. It can be any context and thus doesn't have to wait to re-acquire the same one.

Reply:
Got it. Should it be on every await?

Reply:
For your code, yes. I don't think you have any dependencies on foreground/non-foreground threads being delegated to/from after initially calling something on a specific thread